File: haskell/lifegame.hs

import System.Random
import System.Posix.Unistd (usleep)
import Data.Array
import Data.List.Split (chunksOf)

type Field = Array (Int, Int) Int

main :: IO ()
main = do
  g <- getStdGen -- Gets the global random number generator.
  let height = 20
      width = 40
  lifegame $ initField height width g

lifegame :: Field -> IO ()
lifegame field = do
  clearScreen
  dumpField field
  usleep 100000
  lifegame $ evolve field

initField :: RandomGen g => Int -> Int -> g -> Field
initField height width g = field
  where randomListElem = take (width * height) $ randomRs (0::Int, 1::Int) g
        randomListIndices = [(y,x) | y<-[0..height-1], x<-[0..width-1]]
        randomList = zip randomListIndices randomListElem
        field = array ((0,0),(height-1, width-1)) randomList

evolve :: Field -> Field
evolve field = newField
  where width  = (snd $ snd $ bounds field) + 1
        height = (fst $ snd $ bounds field) + 1
        deadOrAlive field (y,x)
          | aliveNeighbours <= 1 = 0
          | aliveNeighbours == 2 = field ! (y,x)
          | aliveNeighbours == 3 = 1
          | aliveNeighbours >= 4 = 0
          where aliveNeighbours = countAliveNeighbours field (y,x)
        newField = array ((0,0),(height-1, width-1))
                         [((y,x), deadOrAlive field (y,x)) | y<-[0..height-1], x<-[0..width-1]]

countAliveNeighbours :: Field -> (Int, Int) -> Int
countAliveNeighbours field (y,x) = sum neighbours
  where width  = (snd $ snd $ bounds field) + 1
        height = (fst $ snd $ bounds field) + 1
        neighbours = [field ! (mod (y + yi) height, mod (x + xi) width) | yi<-[-1..1], xi<-[-1..1], not (xi == 0 && yi == 0)]

dumpField :: Field -> IO ()
dumpField field = putStrLn $ foldl1 join eachLines
  where width  = (snd $ snd $ bounds field) + 1
        toChar ch = if ch == 0 then ' ' else 'o'
        join a b = a ++ "\n" ++ b
        eachLines = chunksOf width $ map toChar $ elems field

clearScreen :: IO ()
clearScreen = putStrLn "\ESC[;H\ESC[2J"

File: scala/lifegame.scala

object Main {
  def main(args: Array[String]): Unit = {
    var lifegame = new Lifegame(20, 40)
    lifegame.loop()
  }
}

class Lifegame(height: Int, width: Int) {
  val rand = scala.util.Random
  var field = Array.ofDim[Int](height, width)
  for (y <- 0 to height-1; x <- 0 to width-1) {
    field(y)(x) = rand.nextInt(2)
  }

  def loop() = {
    while (true) {
      this.clearScreen
      this.dumpField
      Thread.sleep(100) // 100[ms]
      field = this.evolve
    }
  }

  def evolve(): Array[Array[Int]] = {
    var newField = Array.ofDim[Int](height, width)
    for (y <- 0 to height-1; x <- 0 to width-1) {
      newField(y)(x) =
        countAliveNeighbours(y, x) match {
          case 2 => field(y)(x)
          case 3 => 1
          case _ => 0
        }
    }
    return newField
  }

  def countAliveNeighbours(y: Int, x: Int): Int = {
    val neighbours =
      for {
        yi <- -1 to 1
        xi <- -1 to 1
        if !(yi == 0 && xi == 0)
        y_i = mod(y + yi, height)
        x_i = mod(x + xi, width)
      } yield field(y_i)(x_i)
    return neighbours.sum
  }

  def mod(a: Int, b: Int): Int = {
    if (a > 0) return a % b
    if (a < 0) return (a + b) % b
    return 0
  }

  def dumpField() = {
    for (y <- 0 to height-1) {
      for (x <- 0 to width-1) {
        print(if (field(y)(x) == 0) " " else "o" )
      }
      println()
    }
  }

  def clearScreen() = {
    print("\u001b[;H\u001b[2J")
  }
}
